### PR TITLE
support roialign backward

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

support roialign backward

## How was this patch tested?

unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala [hunk 1]
```diff
@@ -109,8 +111,180 @@ class RoiAlign[T: ClassTag] (
...
+
+      var roi_width = roi_end_w - roi_start_w
+      var roi_height = roi_end_h - roi_start_h
+
+      if (aligned) {
+        require(roi_width >= 0 && roi_height >= 0,
+          s"ROIs in ROIAlignRotated do not have non-negative size!" +
+            s"But get ${roi_height} ${roi_width}")
+      } else {
+        roi_width = math.max(roi_width, 1.0f)
```
Comment (benjamim93) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala]: Why this part is skipped if aligned?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala [hunk 2]
```diff
@@ -109,8 +111,180 @@ class RoiAlign[T: ClassTag] (
...
+              ev.plus(gradInputArr(gradInputOffset + y_high * width + x_high), g4)
+          }
+        }
+      }
+    }
+  }
+
+  override def updateGradInput(input: Activity, gradOutput: Tensor[T]): Activity = {
+    require(mode == "avg", s"Only support backward for average mode, but get ${mode}")
+    val data = input.toTable[Tensor[Float]](1)
```
Comment (benjamim93) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala]: Looks like the backward limited to float only?

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala]: no, it should be classtag `T`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala]: for aligned case, we allow roi height/width less than 1.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala (hunk 2 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala]: Looks good to me.

**Comment by zhangxiaoli73** [comment]
jenkins pass:
http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/1099/

**Pull request merged**